.cursor/plans/ai_feedback_system_design_2afd1713.plan.md

```diff
@@ -0,0 +1,375 @@
+}
+```
+
+### Feedback Categories
+
+| Category Key | Display Name |
+
+|--------------|--------------|
+
+| `instruction_ignored` | Instruction ignored |
+
+| `no_citation_links` | No citation links |
+
+| `being_lazy` | Being lazy |
+
+| `incorrect_information` | Incorrect information |
+
+| `other` | Others |
+
+### Attribution Types
+
+| Attribution | Description | Actionable By |
+
+|-------------|-------------|---------------|
+
+| `rill` | AI made an error in reasoning, tool usage, or response generation | Rill team |
+
+| `project` | Missing or insufficient project data/metadata | Project developer |
+
+### Analytics Query Example
+
+```sql
+SELECT 
+    session_id,
+    created_on,
+    json_extract(content, '$.target_message_id') as target_message_id,
+    json_extract(content, '$.sentiment') as sentiment,
+    json_extract(content, '$.categories') as categories,
+    json_extract(content, '$.comment') as comment
+FROM ai_messages 
+WHERE tool = 'user_feedback' AND type = 'call'
+```
+```sql
+-- Get attribution predictions for negative feedback
+SELECT 
+    m1.session_id,
+    m1.created_on,
+    json_extract(m2.content, '$.predicted_attribution') as attribution,
+    json_extract(m2.content, '$.suggested_action') as suggested_action
+FROM ai_messages m1
+JOIN ai_messages m2 ON m1.parent_id = m2.parent_id 
+    AND m2.type = 'result' AND m2.tool = 'user_feedback'
+WHERE m1.tool = 'user_feedback' AND m1.type = 'call'
+    AND json_extract(m1.content, '$.sentiment') = 'negative'
+```
+
+## API Design
+
+### Extending `CompleteStreaming`
+
+Add a `UserFeedbackContext` parameter to the existing `CompleteStreamingRequest`:
+
+```protobuf
+message CompleteStreamingRequest {
+  string instance_id = 1;
+  string conversation_id = 2;
+  string prompt = 3;
+  string agent = 10;
+  AnalystAgentContext analyst_agent_context = 11;
+  DeveloperAgentContext developer_agent_context = 12;
+  
+  // NEW: Optional feedback context. If provided, records feedback
+  // and returns immediately. Attribution prediction runs async.
+  UserFeedbackContext user_feedback_context = 13;
+}
+
+message UserFeedbackContext {
+  string target_message_id = 1;  // The message being rated
+  string sentiment = 2;          // "positive" or "negative"
+  repeated string categories = 3; // Only for negative sentiment
+  string comment = 4;            // Optional free-text
+}
+```
+
+**Behavior:**
+
+- If `user_feedback_context` is provided: Record feedback call, return immediately (user not blocked)
+- For negative sentiment: Spawn goroutine to run attribution prediction, store result when complete
+
+## Async Attribution Flow
+
+To avoid blocking the user, attribution prediction runs asynchronously:
+
+```
+Frontend submits negative feedback
+    ↓
+CompleteStreaming(user_feedback_context: {...})
+    ↓
+Record user_feedback tool "call" message
+    ↓
```

Comment:
Maybe pass it to the `router_agent` and let it make the `user_feedback` tool call. It'll keep more logic in the `runtime/ai` package (and not in the API handler), and probably make it easier to test.

I also think we might rename `router_agent` to `chat_agent` to reflect that it's really the entrypoint for handling anything from UI chat.